File: 14. Tree/MorrisTraversal.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class MorrisTraversal {
    static List<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        TreeNode curr = root;

        while (curr != null) {
            // Left subtree empty
            if (curr.left == null) {
                result.add(curr.val);
                curr = curr.right;
            }
            // Left subtree exists
            else {
                // Find curr's predecessor(go 1 left and right until null)
                TreeNode prev = curr.left;
                while (prev.right != null && prev.right != curr) {
                    prev = prev.right;
                }
                // Make thread and move curr
                if (prev.right == null) {
                    prev.right = curr;
                    curr = curr.left;
                }
                // Break thread
                else {
                    prev.right = null;
                    result.add(curr.val);
                    curr = curr.right;
                }
            }
        
        }
        return result;
    }
    
    public static void main(String[] args)
    {
        //     1
        //   2    3
        //     5    4

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left.right = new TreeNode(5);
        root.right = new TreeNode(3);
        root.right.right = new TreeNode(4);

        List<Integer> result = inorderTraversal(root);
        System.out.println(result);
    }
}

File: 14. Tree/MorrisTraversal.cpp
//MORRIS TRAVERSAL
//Time Complexity - O(N)
//Space Complexity - O(1)

#include<bits/stdc++.h>
using namespace std;
//Tree Node
struct Node{
  public:
  int data;
  Node* left;
  Node* right;
  Node(int x){
    data=x;
    left=NULL;
    right=NULL;
  }
};
//function defining Morris Traversal
void MorrisTraversal(Node* root){
  Node* curr;
  Node* prev;
  if(root==NULL)return ;
  
  curr=root;
  while(curr!=NULL){
    if(curr->left==NULL){
      cout<<curr->data<<" ";
      curr=curr->right;
    }
    else{
      prev=curr->left;
      while(prev->right!=NULL && prev->right!=curr){
        prev=prev->right;
      }
      if(prev->right==NULL){
        prev->right=curr;
        curr=curr->left;
      }  
      else{
        prev->right=NULL;
        cout<<curr->data<<" ";
        curr=curr->right;
      }  
    }  
  }  
}  
int main(){
  Node* root=new Node(10);
  root->left=new Node(20);
  root->right=new Node(30);
  root->left->left=new Node(40);
  root->left->right=new Node(50);
  root->right->left=new Node(60);
  root->right->right=new Node(70);
  MorrisTraversal(root);
  return 0;
}
